include/xgboost/data.h

```diff
@@ -260,28 +313,17 @@ class DMatrix {
    * \brief get the row iterator, reset to beginning position
    * \note Only either RowIterator or  column Iterator can be active.
    */
-  virtual dmlc::DataIter<RowBatch>* RowIterator() = 0;
+  virtual dmlc::DataIter<SparsePage>* RowIterator() = 0;
   /*!\brief get column iterator, reset to the beginning position */
-  virtual dmlc::DataIter<ColBatch>* ColIterator() = 0;
-  /*!
-   * \brief get the column iterator associated with subset of column features.
-   * \param fset is the list of column index set that must be contained in the returning Column iterator
-   * \return the column iterator, initialized so that it reads the elements in fset
-   */
-  virtual dmlc::DataIter<ColBatch>* ColIterator(const std::vector<bst_uint>& fset) = 0;
```

Comment:
Yes something like that.

Consider the old and new syntax:
```c++
// Old
 std::vector<int> fset;
  auto iter = p_fmat->ColIterator(fset);
  while (iter->Next()) {
    auto batch = iter->Value();
    for(int i = 0; i < batch.Size(); i++)
    {
      auto col = batch[i];
      auto fidx = batch.col_index[i];
    }
  }

// New
  std::vector<int> fset;
  auto iter = p_fmat->ColIterator();
  while (iter->Next()) {
    auto batch = iter->Value();
    for (auto fidx : fset)
    {
      auto col = batch[fidx];
    }
  }
```

The new code is shorter by one line and provides certainty about the state of batch because it only ever contains all of the features. In short it seems easier and safer to let the user locally manage the set of features they are interested in. Batch is now a much simpler object.

Reply:
In other words, we achieved simplicity at no added computational cost. Cool!